CalPatRec/src/HelixFinder_module.cc

```diff
@@ -0,0 +1,2559 @@
+    fhicl::Atom<float> minHelixPerpMomentum{Name("minHelixPerpMomentum"),
+                                            Comment("min pt of helix")};
+    fhicl::Atom<float> maxHelixPerpMomentum{Name("maxHelixPerpMomentum"),
+                                            Comment("max pt of helix")};
+    fhicl::Atom<float> minHelixMomentum{Name("minHelixMomentum"), Comment("min momentum of helix")};
+    fhicl::Atom<float> maxHelixMomentum{Name("maxHelixMomentum"), Comment("max momentum of helix")};
+    fhicl::Atom<float> chi2LineSaveThresh{Name("chi2LineSaveThresh"),
+                                          Comment("max chi2Dof for line")};
+    fhicl::Atom<int> debugPdgID{Name("debugPdgID"), Comment("pdgID of interest in display")};
+    fhicl::Atom<float> debugMomentum{Name("debugMomentum"), Comment("lower momentum of interest")};
+    fhicl::Atom<std::string> debugDirection{Name("debugDirection"), Comment("down or up")};
+    fhicl::Atom<int> debugStrawHitThresh{Name("debugStrawHitThresh"),
+                                         Comment("min #SHs for plot particle")};
+    fhicl::Atom<float> debugScatterThresh{Name("debugScatterThresh"),
+                                          Comment("max dP for plot particle")};
+    fhicl::Table<McUtilsToolBase::Config> mcUtils{Name("mcUtils"),
+                                                  Comment("get MC info if debugging")};
+    fhicl::Table<HelixFinderTypes::Config> diagPlugin{Name("diagPlugin"), Comment("diag plugin")};
+  };
+
+  struct cHit {
+    int hitIndice; // index of point in _chColl
+    float circleError2;
+    float helixPhi;
+    float helixPhiError2;
+    int helixPhiCorrection;
+    bool inHelix;
+    bool used; // whether or not hit is used in fits
+    bool isolated;
+    bool averagedOut;
+    bool notOnLine;
+    bool uselessTripletSeed;
+    bool notOnSegment;
+    bool debugParticle; // only filled in debug mode -- true if mc particle, false if background
+  };
+
+  struct tripletPoint {
+    const XYZVectorF* pos;
+    int hitIndice;
+  };
+
+  struct triplet {
+    tripletPoint i;
+    tripletPoint j;
+    tripletPoint k;
+  };
+
+  struct lineInfo {
+    float zMin;
+    float zMax;
+    ::LsqSums2 fitter;
+    std::vector<size_t> tcHitsIndices;
+    std::vector<int> helixPhiCorrections; // integer for 2pi ambiguity
+  };
+
+  struct helixCandidate {
+    std::vector<cHit> tcHitsCopy;
+    ::LsqSums4 circleFitterCopy;
+    ::LsqSums2 lineFitterCopy;
+  };
+
+  // struct to hold info specifically for when debugging
+  struct logicTrackingInfo {
+    std::vector<std::vector<cHit>> tcHitsColl;
+    std::vector<::LsqSums4> circleFitter;
+    std::vector<::LsqSums2> lineFitter;
+    std::vector<lineInfo> lineSegments;
+    std::vector<lineInfo> resolvedSegments;
+    bool foundLineSegment;
+    size_t bestLineSegment;
+    int finalNHits;
+    bool becameHelix;
+  };
+
+  // another struct for debugging
```

Comment:
Hey Dave, this is understandable. I wanted to avoid this and eventually remove it completely (which is why I held off committing my code as long as I did). The MC stuff is solely for debugging. I wanted originally to have it contained in a separate diagnostic module, but determined that this wouldn't be ideal ... decisions are made in the debug plotting that have to be made as the helix is being built (in debug mode a plot is made at every stage of the helix reconstruction for the helix that made it the furthest in the reconstruction for the particle of interest ... so you can see the stage that caused failure ... it also color coats the hits so you can see bkg/signal), and to make those decisions in a separate module would require storing a LOT of collections (for example you have many triplet candidates for each time cluster, and then a lot of time clusters quite often and without the MC info you don't know which one contains stuff you actually want to debug/plot). For development purposes I decided then to keep the MC tools in the module and eventually remove them when I was done developing. So maybe we can allow it for now and remove it in a few months time? If not, I can remove it now. I can understand not wanting MC tools in producer modules in the Offline.